custom paths in settings schema

shahzebsiddiqui opened the pull request:
Summary
-----------

- Adding option `buildtest show schema --settings` to show settings schema
- Add `config` section in settings schema with subsection `paths` where we can define the test directory, location where to search buildspecs which is a colon separate list that mimics $PATH but for buildspecs, and location where to clone repos via `buildtest get`.

shahzebsiddiqui committed:
adding option buildtest show schema --settings with short option -s for displaying the
settings schema.
Adding test for coverage for this option

shahzebsiddiqui committed:
update settings schema with new section 'config' where users can define
buildtest configuration, we have a section 'path' under config where
we can define location where to clone repo via 'buildtest get' using key
get_repo_path.
You can configure path where tests can be written via 'testdir'. The precendence
is given to command line, then settings file, and lastly current directory.
The buildspec_path is a colon separated list to act like $PATH for buildspecs, this
adds extra paths where user can specify a buildspec which works well when you can clone a repo
in arbitrary location.

shahzebsiddiqui committed:
Merge branch 'devel' into settings_rev2

shahzebsiddiqui committed:
applying black

shahzebsiddiqui committed:
fix bug when processing nested dictionary. For now user are not required
to define the 'config' section in schema so we introduced a bug in
previous commit which this attempts to solve. The bug was related
present in buildtest build and buildtest get

shahzebsiddiqui committed:
update urls for HPCKP17,18 reported by urlcheckers. Removed Powerpoint slides,
its sufficient to have PDF copies as reference.

vsoch commented on buildtest/menu/build.py:
```diff
@@ -18,7 +18,7 @@
 
 
-def discover_buildspecs(buildspec):
+def discover_buildspecs(buildspec, search_path=[]):
```
You should generally not set a list as an input argument to a function, it's mutable. Instead this should be set to None, and then `search_path = search_path or []`

vsoch commented on buildtest/menu/build.py:
```diff
@@ -38,15 +38,17 @@ def discover_buildspecs(buildspec):
     # Second try, the path can be relative to the BUILDSPEC_DEFAULT_PATH
-    elif not os.path.exists(buildspec):
-        buildspec = os.path.join(BUILDSPEC_DEFAULT_PATH, buildspec)
+    if not os.path.exists(buildspec):
```
Why did you remove the check here if it's set to None? This is again going to trigger an error if this is the case.

shahzebsiddiqui replied:
Well i don't anticipate user will pass `buildtest build` we shouldnt even allow this. They should pass with `-b` option. That's why i removed the check

vsoch commented on buildtest/menu/build.py:
```diff
@@ -64,9 +66,10 @@ def discover_buildspecs(buildspec):
-            "Please provide an absolute or relative path to a directory file from your present working directory or %s"
-            % BUILDSPEC_DEFAULT_PATH
+            f"Unable to find any buildspecs in search paths: {search_path} \n"
+            + "Please provide an absolute or relative path to a directory or file or file relative to current directory."
```
or file or file

vsoch commented on buildtest/menu/build.py:
```diff
@@ -148,13 +151,32 @@ def func_build_subcmd(args):
+    if config_paths_buildspec_path:
+        # return a unique list of directory paths where buildtest should search for buildspecs
+        # this is equivalent to $PATH for buildspec
+        buildspec_paths = list(set(config_paths_buildspec_path.split(":")))
```
No need to strip for empty spaces (e.g., if there is a trailing : you are going to get an empty string as a path which will have unintended consequences...

shahzebsiddiqui replied:
I see the trailing `:` is an issue i can put a check and remove it if found at end of char, we okay with splitting by colon i assume since this check would be done by default.

shahzebsiddiqui replied:
i tested this locally and got this resolved, see latest commit

shahzebsiddiqui committed:
rename keys in schema to searchpath and clonepath
Fix bug when trailing colon specified. Also take into account if no
colon is specified for searchpath.
